query/sortUser.go

```diff
@@ -0,0 +1,103 @@
+package query
+
+import (
+	"github.com/ONSdigital/dp-identity-api/models"
+	"sort"
+	"strings"
+)
+
+type lessFunc func(p1, p2 *models.UserParams) bool
+
+type multiSorter struct {
+	changes []models.UserParams
+	less    []lessFunc
+}
+
+func (ms *multiSorter) Sort(changes []models.UserParams) {
+	ms.changes = changes
+	sort.Sort(ms)
+}
+
+func OrderedBy(less ...lessFunc) *multiSorter {
+	return &multiSorter{
+		less: less,
+	}
+}
+func (ms *multiSorter) Len() int {
+	return len(ms.changes)
+}
+func (ms *multiSorter) Swap(i, j int) {
+	ms.changes[i], ms.changes[j] = ms.changes[j], ms.changes[i]
+}
+func (ms *multiSorter) Less(i, j int) bool {
+	p, q := &ms.changes[i], &ms.changes[j]
+	var k int
+	for k = 0; k < len(ms.less)-1; k++ {
+		less := ms.less[k]
+		switch {
+		case less(p, q):
+			return true
+		case less(q, p):
+			return false
+		}
+	}
+	return ms.less[k](p, q)
+}
+
+func SortBy(paramsSlice string, arr []models.UserParams) {
+	forenameAsc := func(c1, c2 *models.UserParams) bool { return c1.Forename < c2.Forename }
+	forenameDesc := func(c1, c2 *models.UserParams) bool { return c1.Forename > c2.Forename }
+	lastnameAsc := func(c1, c2 *models.UserParams) bool { return c1.Lastname < c2.Lastname }
+	lastnameDesc := func(c1, c2 *models.UserParams) bool { return c1.Lastname > c2.Lastname }
+	emailAsc := func(c1, c2 *models.UserParams) bool { return c1.Email < c2.Email }
+	emailDesc := func(c1, c2 *models.UserParams) bool { return c1.Email > c2.Email }
+	idAsc := func(c1, c2 *models.UserParams) bool { return c1.ID < c2.ID }
+	idDesc := func(c1, c2 *models.UserParams) bool { return c1.ID > c2.ID }
```

Comment:
Could this be made more generic by passing attributes in? i.e. we could map the incoming attributes to the underlying struct variables. Covers a bit of those nested if statements later as well.